contracts/Ethernauts.sol

```diff
@@ -75,23 +114,48 @@ contract Ethernauts is ERC721Enumerable, Ownable {
         mintPrice = newMintPrice;
     }
 
-    function setBaseURI(string memory baseTokenURI_) public onlyOwner {
-        baseTokenURI = baseTokenURI_;
+    function setEarlyMintPrice(uint newEarlyMintPrice) external onlyOwner {
+        earlyMintPrice = newEarlyMintPrice;
+    }
+
+    function setBaseURI(string memory newBaseTokenURI) external onlyOwner {
+        baseTokenURI = newBaseTokenURI;
+    }
+
+    function setSaleState(SaleState newSaleState) external onlyOwner {
+        require(newSaleState != currentSaleState, "Invalid new state");
+
+        currentSaleState = newSaleState;
+    }
+
+    function setCouponSigner(address newCouponSigner) external onlyOwner {
+        couponSigner = newCouponSigner;
     }
 
     function withdraw(address payable beneficiary) external onlyOwner {
         beneficiary.sendValue(address(this).balance);
     }
 
     // -------------------
-    // Internal functions
+    // Private functions
     // -------------------
 
+    function _verifyCoupon(bytes memory signedCoupon) private {
+        bytes32 messageHash = keccak256(abi.encode(msg.sender));
+        require(!_couponUsed[msg.sender], "Expired coupon");
+
+        bytes32 prefixedHash = ECDSA.toEthSignedMessageHash(messageHash);
```

Comment:
nit: would be cleaner if this precondition was checked in the first line of the function

```suggestion
    function _verifyCoupon(bytes memory signedCoupon) private {
        require(!_couponUsed[msg.sender], "Expired coupon");
        
        bytes32 messageHash = keccak256(abi.encode(msg.sender));
        bytes32 prefixedHash = ECDSA.toEthSignedMessageHash(messageHash);
```